File: javascript/15-valid_anagram.js
// # Given two strings s and t, return true if t is an anagram of s, and false otherwise.

// # An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

// # Example 1:

// # Input: s = "anagram", t = "nagaram"
// # Output: true
// # Example 2:

// # Input: s = "rat", t = "car"
// # Output: false

const validAnagram = (s, t) => {
  if (s.length !== t.length) return false;

  for (let i = 0; i < t.length; i++) {
    if (!s.includes(t[i])) return false;

    let loc = s.indexOf(t[i]);
    s = s.slice(0, loc) + s.slice(loc + 1);
  }
  return true;
};

// 2nd Approach
const validAnagram2 = (s, t) => {
  if (s.length !== t.length) return false;

  let common = new Map();
  for (let i = 0; i < s.length; i++) {
    s[i] in common ? common[s[i]]++ : (common[s[i]] = 1);
  }

  for (let j = 0; j < t.length; j++) {
    if (t[j] in common && common[t[j]] >= 1) {
      common[t[j]]--;
    } else {
      return false;
    }
  }
  return true;
};

s = "anagram";
t = "nagaram";

console.log(validAnagram2(s, t));

File: python/15-valid_anagram.py
# Given two strings s and t, return true if t is an anagram of s, and false otherwise.

# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.


# Example 1:

# Input: s = "anagram", t = "nagaram"
# Output: true
# Example 2:

# Input: s = "rat", t = "car"
# Output: false

def validAnagram(s: str, t: str):
    if (len(s) != len(t)):
        return False

    map1 = {}
    for ch in s:
        map1[ch] = map1.get(ch, 0) + 1
    for char in t:
        if (char in map1.keys() and map1[char] >= 1):
            print(map1)
            map1[char] -= 1
        else:
            return False
    return True

# 2nd Approach


def validAnagram2(s: str, t: str):
    if (len(s) != len(t)):
        return False

    for ch in t:
        if ch not in s:
            return False

        location = s.index(ch)
        s = s[0:location] + s[location + 1:]

    return True


# s = "anagram"
# t = "nagaram"
s = "rat"
t = "car"
print(validAnagram2(s, t))
